### DISCUSSION TITLE
Part 2 negative values

### DISCUSSION THREAD

**Discussion by AdamG412** [wenceorg/comp4187#97] [Q&A]
For the Darcy equation, I've been having an issue  where the top half of my solution is positive and the bottom is negative. For example on a 4x4x4 grid my solution is:
 [[ 0.          0.          0.          0.        ]
  [ 0.          0.82210677  0.82210677  0.        ]
  [-0.         -0.82210677 -0.82210677 -0.        ]
  [-0.         -0.         -0.         -0.        ]]

 [[ 0.          0.          0.          0.        ]
  [ 0.          0.82210677  0.82210677  0.        ]
  [-0.         -0.82210677 -0.82210677 -0.        ]
  [-0.         -0.         -0.         -0.        ]]
 (for the middle 2 y levels, the rest are 0) and the exact solution is:
 [[0.         0.         0.         0.        ]
  [0.         0.64951905 0.64951905 0.        ]
  [0.         0.64951905 0.64951905 0.        ]
  [0.         0.         0.         0.        ]]

 [[0.         0.         0.         0.        ]
  [0.         0.64951905 0.64951905 0.        ]
  [0.         0.64951905 0.64951905 0.        ]
  [0.         0.         0.         0.        ]]
Since we have cos(pi*y) in the left hand side it makes sense that this would be the case, as for y>0.5 this term is negative. I've tested my stencil with k=1 as you suggested in a previous answer, and this seems to work, but I'm having trouble pinpointing where in the stencil I'm going wrong, either with the cos(pi * y) part or the part involving the derivative of k.

**Comment by annereinarz** [comment]
On a grid of this size you might get some larger errors, since you're essentially interpolating cos(pi y) and it actually varies a fair amount over the domain.
That said, there does appear to be a sign error somewhere in you calculation. You are working on [0,1]^3 though, right? Not [-1,1]? If you'd missed the sign in the derivative I expect you'd have this issue on both sides (wrong sign). If you've got a shift (+pi/2) in the cos/sin that might give you this result though.
